### DISCUSSION TITLE
CGAL Shape Regularization

### DISCUSSION THREAD

**Discussion by luozi123456** [CGAL/cgal#6451] [Q&A]
I want to run the demo named "Angle + Offset Regularization" .But my VS2017 kept report error "C2947" point to the OCtree.h of CGAL.Mean while the code "using Quadratic_program =
CGAL::OSQP_quadratic_program_traits<FT>;" has been labeled,and show "CGAL has no member of  OSQP_quadratic_program_traits".
Once i cancel include "Shape_regularization.h",the C2947 just disply.
Dose anyone know how to figure this problem?
![image](https://user-images.githubusercontent.com/58937480/160137381-a867f0ef-91a9-46ba-b7d5-95156782162c.png)

**Comment by lrineau** [comment]
Have you used CMake to produce the solution file for VS2017?

I see three possibilities that would explain why `CGAL::OSQP_quadratic_program_traits` does not exist:
  - Maybe you do not have OSQP installed.
  - Or CMake was not able to find your installation of OSQP.
  - Or you have not generated the solution with CMake, and the preprocessor macro `CGAL_USE_OSQP` was not defined.


I note that you have send me a personal email, and posted here in Discussions at the same time. Please do not send personal emails, and only post here.

**Reply by luozi123456** [reply]
Thanks for your advice.I've solved the problem that QSQP does not exit according to your suggestions.But there still some bugs.
As you can see,there are 6 errors as followed 
C2941  should by "& gt;"Terminates the template parameter list but finds "&lt;"CGAL_TEST
C2365 Gt: redefine.The previous definition was "template parameters"
C2988 Unrecognized template declares/defines CGAL_TEST
C2059 Syntax error: ", "CGAL_TEST
C2143 Syntax error: Missing '; '(before '{') CGAL_TEST
C2447 "{" : Missing function title (old fashioned form table?)CGAL_TEST

![image](https://user-images.githubusercontent.com/58937480/161991024-e2f60aba-e6c2-4742-8bcc-ac0a06a0ac28.png)

I think it's line 46-47 in the "octree.h" file or Shape_regularization.h ".
Because only section 2.4 of the GDAL-Shape_regularization case has this error, the rest of the cases work fine.Could you give me some advice on this problem?

**Reply by lrineau** [reply]
Which version of CGAL do you use?

**Reply by luozi123456** [reply]
CGAL-5.4

**Reply by lrineau** [reply]
I do not understand the messages from the compiler. They make no sense. My guess would be that there is a macro definition, somewhere, that redefines `IR`, or `IPM`, or `INM` to something the compiler cannot parse correctly. Search for `#define IR` in your code, for example. Unfortunately, I do not think we can help more. Try to split your program into smaller parts, to see if you can make that compile if you remove (or comment) part of the code.

**Reply by luozi123456** [reply]
Thanks for your reply.
